src/hotspot/share/opto/loopUnswitch.cpp

```diff
@@ -343,6 +407,171 @@ void PhaseIdealLoop::do_unswitching(IdealLoopTree* loop, Node_List& old_new) {
+//   multiversion_fast_proj          multiversion_slow_proj
+//                                                |
+//                                                +--------+
+//                                                         |
+//                                                      slow_path
+//
+//
+// After:
+//                     entry  opaque <-- to be replaced by caller
+//                         |  |
+//                        new_if
+//                         |  |
+//                         |  +-----------------------------+
+//                         |                                |
+//                 new_if_true  opaque                new_if_false
+//                         |      |                         |
+//                      multiversion_if                     |
+//                         |      |                         |
+//        +----------------+      +---------------+         |
+//        |                                       |         |
+//   multiversion_fast_proj      new_multiversion_slow_proj |
+//                                                |         |
+//                                                +------+  |
+//                                                       |  |
+//                                                      region
+//                                                         |
+//                                                      slow_path
+//
+IfTrueNode* PhaseIdealLoop::create_new_if_for_multiversion(IfTrueNode* multiversioning_fast_proj) {
+  // Give all nodes in the old sub-graph a name.
+  IfNode* multiversion_if = multiversioning_fast_proj->in(0)->as_If();
+  Node* entry = multiversion_if->in(0);
+  OpaqueMultiversioningNode* opaque = multiversion_if->in(1)->as_OpaqueMultiversioning();
+  IfFalseNode* multiversion_slow_proj = multiversion_if->proj_out(0)->as_IfFalse();
+  Node* slow_path = multiversion_slow_proj->unique_ctrl_out();
+
+  // Now that we have at least one condition for the multiversioning,
+  // we should unstall the slow loop.
+  opaque->unstall_slow_loop();
+
+  // Create new_if with its projections.
+  IfNode* new_if = IfNode::make_with_same_profile(multiversion_if, entry, opaque);
+  IdealLoopTree* lp = get_loop(entry);
+  register_control(new_if, lp, entry);
+
+  IfTrueNode*  new_if_true  = new IfTrueNode(new_if);
+  IfFalseNode* new_if_false = new IfFalseNode(new_if);
+  register_control(new_if_true,  lp, new_if);
+  register_control(new_if_false, lp, new_if);
+
+  // Hook new_if_true into multiversion_if.
+  _igvn.replace_input_of(multiversion_if, 0, new_if_true);
+
+  // Clone multiversion_slow_path - this allows us to easily carry the dependencies to
+  // the new region below.
+  IfFalseNode* new_multiversion_slow_proj = multiversion_slow_proj->clone()->as_IfFalse();
+  register_control(new_multiversion_slow_proj, lp, multiversion_if);
+
+  // Create new Region.
+  RegionNode* region = new RegionNode(1);
```

Comment:
So we create a new `Region` every time a new condition is added?

Reply:
Yes. Are you ok with that? Or would you prefer if we extended an existing region (is that possible?) and then we'd have 2 cases, one where there is none yet, and one where we'd extend. I think adding one each time is easier, and it would get commoned anyway, right?

Reply:
That sounds ok to me.